Implement access=user on FreeBSD using nfs-ganesha-kmod (Closes chaos/diod#55)

lfourquaux opened the pull request:
This is an attempt at implementing access=user on FreeBSD using the syscalls provided by nfs-ganesha-kmod.

garlick commented:
Wow, cool!

I think this is probably too invasive to mix together with the existing user.c code, which is already hard to read.   It's dangerous for me to let that change too much, since I don't have a lot of time to invest in this project at the moment.  Changing it is a bit scary from a security point of view.  (I'm wondering now if that capability should be disabled by default even on Linux actually)

Perhaps we can factor out the non-portable bits of both Linux and FreeBSD into separate source module(s)?

There should probably be an opt-in autoconf `--with-ganesha-kmod`  option or similar, to avoid assuming that all freebsd users would have that module available.

Re: the standalone portability fixes, see comment in other PR about using autoconf tests rather than `#ifdef __FreeBSD__`.  If you want to tack those onto that PR to get them in early, I have no objection (even though there's a short circuit out of the user code for freebsd at present).

garlick commented:
>  If you want to tack those onto that PR to get them in early, I have no objection (even though there's a short circuit out of the user code for freebsd at present).

Or put those in another PR that can be more quickly reviewed and merged, since that one referenced was merged already.

lfourquaux commented:
I've updated this pull request to leave it in a clean state while I try to do more portability fixes and split the non-portable bits out of user.c.

lfourquaux commented:
I've updated the pull request regarding FreeBSD:
 * the ganesha-kmod code is disabled by default, use --enable-impersonation to build it;
 * if the kernel modules are not loaded, changing users will fail but running as a single user will work as before;
 * I have bundled some small cleanups in this pull request (see first commit), I hope this is not a problem (I can submit a separate pull request if necessary);
 * as a side effect of these changes, --disable-impersonation also disables some Linux-specific code (that is irrelevant if user-stub.c is used);
 * there is the usual git dance to make a copy of user-linux.c while preserving history;
 * I have added one header file and I don't care what license is used for these few lines of code, so let's say it's in the public domain.

lfourquaux commented on liblsd/hostlist.c:
```diff
@@ -381,7 +381,7 @@ static char * _next_tok(char *sep, char **str)
 
     /* push str past any leading separators */
-    while (**str != '\0' && strchr(sep, **str) != '\0')
+    while (**str != '\0' && strchr(sep, **str) != NULL)
```
Do you want only hostlist.c, or the whole first commit?

garlick replied:
All the changes in this commit are good - just wanted to split it into two commits, one that applies the changes to hostlist.c.  (We use that module in other projects so it would be nice to be able to refer to a standalone commit that improves it).

Thanks!

lfourquaux commented:
Thanks!  I've split the first commit, and (minimally) documented the changes.  Is this OK?

garlick commented:
Ready to merge?

lfourquaux commented:
Yes, this should be ready to merge.  Thanks again!

garlick commented:
I'll need to do something about the broken test suite before we make the next tag.  I need to allocate some time for that, so it may be a while, but I'll try to get it scheduled.
